test/language/default-parameters/call-define-values.js

```diff
@@ -0,0 +1,35 @@
+// Copyright (C) 2015 the V8 project authors. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+es6id: 13.3.3.7
+description: >
+  Function call define default values to arguments
+info: >
+  13.3.3.7 Runtime Semantics: KeyedBindingInitialization
+
+  When undefined is passed for environment it indicates that a PutValue
+  operation should be used to assign the initialization value. This is the case
+  for formal parameter lists of non-strict functions. In that case the formal
+  parameter bindings are preinitialized in order to deal with the possibility of
+  multiple parameters with the same name.
+---*/
+
+var results;
+
+var o = {};
+
+function fn(a = 1, b = null, c = o, d) {
+  return [a, b, c, d];
+}
+
+results = fn();
+
+assert.sameValue(results[0], 1, 'apply default values #1');
+assert.sameValue(results[1], null, 'apply default values #2');
+assert.sameValue(results[2], o, 'apply default values #3');
+
+// Ref: https://bugzilla.mozilla.org/show_bug.cgi?id=777060
```

Comment:
I've referenced Mozilla's BugZilla tracker for many personal projects, but
    because these are the canonical ECMAScript tests, I think we should derive
    all justification from the spec itself. You were right about the spec being
    tricky here; I think the justification goes something like this:

    > 24. If hasDuplicates is true, then
    >     [...]
    > 25. Else,
    >     a. Let formalStatus be IteratorBindingInitialization for formals with
    >        iteratorRecord and env as arguments.

    > ES6 13.3.3.6 Runtime Semantics: IteratorBindingInitialization
    >
    > SingleNameBinding : BindingIdentifier Initializeropt
    >
    > [...]
    > 6. If Initializer is present and v is undefined, then
    >    [...]

Seeing that justification makes an improvement more obvious: this test is
    good because it ensures that the default value is not used when a
    non-undefined positional argument is specified. It would be even better if
    it ensured that the initializer is not evaluated in those cases.